google/cloud/bigquery/v2/minimal/internal/job_request.cc

```diff
@@ -300,62 +300,83 @@ void to_json(nlohmann::json& j, QueryRequest const& q) {
 }
 
 void from_json(nlohmann::json const& j, PostQueryRequest& q) {
-  if (j.contains("projectId")) {
-    q.set_project_id(j.at("projectId").value("projectId", ""));
-  }
-  if (j.contains("queryRequest")) {
-    q.set_query_request(j.at("queryRequest").get<QueryRequest>());
-  }
+  std::string project_id;
+  SafeGetTo(project_id, j, "projectId");
+  q.set_project_id(project_id);
+
+  QueryRequest query_request;
+  SafeGetTo(query_request, j, "queryRequest");
+  q.set_query_request(query_request);
 }
 
 void from_json(nlohmann::json const& j, QueryRequest& q) {
-  if (j.contains("query")) q.set_query(j.at("query").get<std::string>());
-  if (j.contains("kind")) q.set_kind(j.at("kind").get<std::string>());
-  if (j.contains("parameterMode")) {
-    q.set_parameter_mode(j.at("parameterMode").get<std::string>());
-  }
-  if (j.contains("location")) {
-    q.set_location(j.at("location").get<std::string>());
-  }
-  if (j.contains("requestId")) {
-    q.set_request_id(j.at("requestId").get<std::string>());
-  }
-  if (j.contains("dryRun")) q.set_dry_run(j.at("dryRun").get<bool>());
-  if (j.contains("preserveNulls")) {
-    q.set_preserve_nulls(j.at("preserveNulls").get<bool>());
-  }
-  if (j.contains("useQueryCache")) {
-    q.set_use_query_cache(j.at("useQueryCache").get<bool>());
-  }
-  if (j.contains("useLegacySql")) {
-    q.set_use_legacy_sql(j.at("useLegacySql").get<bool>());
-  }
-  if (j.contains("createSession")) {
-    q.set_create_session(j.at("createSession").get<bool>());
-  }
-  if (j.contains("maxResults")) {
-    q.set_max_results(j.at("maxResults").get<std::uint32_t>());
-  }
-  if (j.contains("maximumBytesBilled")) {
-    q.set_maximum_bytes_billed(j.at("maximumBytesBilled").get<std::int64_t>());
-  }
-  if (j.contains("connectionProperties")) {
-    q.set_connection_properties(
-        j.at("connectionProperties").get<std::vector<ConnectionProperty>>());
-  }
-  if (j.contains("queryParameters")) {
-    q.set_query_parameters(
-        j.at("queryParameters").get<std::vector<QueryParameter>>());
-  }
-  if (j.contains("defaultDataset")) {
-    q.set_default_dataset(j.at("defaultDataset").get<DatasetReference>());
-  }
-  if (j.contains("formatOptions")) {
-    q.set_format_options(j.at("formatOptions").get<DataFormatOptions>());
-  }
-  if (j.contains("labels")) {
-    q.set_labels(j.at("labels").get<std::map<std::string, std::string>>());
-  }
+  std::string query;
+  SafeGetTo(query, j, "query");
+  q.set_query(query);
+
+  std::string kind;
+  SafeGetTo(kind, j, "kind");
+  q.set_kind(kind);
+
+  std::string parameter_mode;
+  SafeGetTo(parameter_mode, j, "parameterMode");
+  q.set_parameter_mode(parameter_mode);
+
+  std::string location;
+  SafeGetTo(location, j, "location");
+  q.set_location(location);
+
+  std::string request_id;
+  SafeGetTo(request_id, j, "requestId");
+  q.set_request_id(request_id);
+
+  bool dry_run;
```

Comment:
Oops I am just noticing that you need to initialize these variables.

`SafeGetTo` does not always set their value.

Reply:
I am not sure that is the "Right Thing":tm: Previously the code looked like this:

```cc
if (j.contains("dryRun")) q.set_dry_run(j.at("dryRun").get<bool>());
```

If we unroll the functions, now it looks more like this:

```cc
bool dry_run;
if (j.contains("dryRun")) dry_run = j.at("dryRun").get<bool>();
q.set_dry_run(dry_run);
```

I think we want something like this:

```cc
auto if_present = [&j](char const* key, auto const& apply) {
  if (auto i = j.at(key); i != j.end()) apply(*i); // TODO: map to C++14, this requires C++17
};

if_present("dryRun", [&q](auto const& j) { q.set_dry_run(j.get<bool>()); });
if_present("preserveNulls", [&q](auto const& j) { q.set_preserve_nulls(j.get<bool>()); }); 
... ...
```

https://godbolt.org/z/dEqqxrzh6